### ISSUE TITLE
ServerSide Autorun Triggers 'Removed' Events Bug

### ISSUE THREAD

**Issue by elm2178** [cult-of-coders/redis-oplog#281]
We use serverside autoruns pretty heavily in our codebase. Upon adding a new user to our workspace model we sometimes see all the existing users get removed from the client side collection. This does not happen every time so that leads me to believe there is a race condition that is causing this. Looking at the ddp messages sent through websocket, I noticed that the server is actually sending 'removed' events down to the client. I was not able to reproduce this exactly in this simple repo but I think I was able to get at the core of the problem.

Here is a simple reproduction repo:
https://github.com/hiveteams/redis-oplog-reproduction

## Issue
When changes to the workspace document are made (ex: a new user being added to the list of workspace members), the redis oplog causes removed events to be propagated to the client resulting in users disappearing from the client collection.

## Steps to reproduce
1. Run `meteor npm i`
2. Start the server with `npm start`
3. Load localhost:3000
4. Make sure redis oplog is enabled
5. Click to add 100 users (might have to do this more than once)
6. Notice that the users are not cleanly added, several 'removed' events are sent through ddp.

If you toggle to disable redis oplog and then try again you will notice that no 'removed' events are sent through ddp.

## Workspace
```json
{
  "_id": "Mongo id",
  "members": "List of user ids"
}
```

## User
```json
{
  "_id": "Mongo id"
}
```

## The Code

Basically just a publish where we depend on the results of the workspaces query to return the final set of users:

```javascript
Meteor.publish('allUsers', function allUsersPub(enableRedisOplog) {
  this.autorun(() => {
    // listen to changes in the workspace document
    const workspace = Workspaces.find().fetch()[0];

    // optionally disable redis oplog
    const options = {};
    if (!enableRedisOplog) {
      options.disableOplog = true;
    }

    // return users cursor
    return Users.find({
      _id: { $in: workspace.members }
    }, options);
  });
});
```

**Comment by etyp** [comment]
Uses [peerlibrary:server-autorun](https://github.com/peerlibrary/meteor-server-autorun) + [peerlibrary:reactive-publish](https://github.com/peerlibrary/meteor-reactive-publish)

**Comment by ramezrafla** [comment]
Is this different from #276?

**Comment by etyp** [comment]
@ramezrafla might be same underlying cause, but the result is different reread because the user documents are never meant to be removed (they always match publication selector). We filed separately because I don't believe this is solved by just swallowing the error meteor throws. The removed messages shouldn't be firing at all here.

**Comment by elm2178** [comment]
I also created a branch 'bug/simple-autorun-not-working' that shows how when redis-oplog is enabled, serverside autoruns do not rerun as expected. Here is the link to that branch:
https://github.com/hiveteams/redis-oplog-reproduction/tree/bug/simple-autorun-not-working

**Comment by ramezrafla** [comment]
Thanks @etyp 

@theodorDiaconu do you have bandwidth? From experience this issue could be complex to trace due to reactivity. Would you accept a bounty to look at it? (Our small way to say thank you). For our deployment, that is the last outstanding issue for redis-oplog

BTW, we see less of it now that we moved mongodb to a cluster that is further away. Maybe latency has something to do with it / race conditions.

**Comment by theodorDiaconu** [comment]
@ramezrafla @elm2178 @etyp this issue is on my today's queue, now that I have reproduction it should be easier to solve, but I have no clue about the estimate. If I get in the zone and things go smoothly we'll do it.

**Comment by theodorDiaconu** [comment]
Fixed. Please read: https://github.com/cult-of-coders/redis-oplog/pull/283

**Comment by ramezrafla** [comment]
Can we close?
